### PR TITLE
HardwareReport: Add OpenAI Assistant to Hardware Report Tool (Chat + Action Support)

### PR DESCRIPTION
**Pull request by AnasChebili**
This PR adds an AI assistant chat interface to the Hardware Report Tool as part of the broader effort to enhance ArduPilot's web tools with intelligent assistant capabilities.

The assistant is capable of answering general questions about the webtool, answering questions about specific user-uploaded logs files, and performing predefined actions.

Currently, this requires using an OpenAI API key to use.

Below is a video showcasing a typical interaction:

https://github.com/user-attachments/assets/f669f173-67aa-424c-b423-fc00811d5c35

Assistant Capabilities:

- Answers general questions about the tool from its instruction set.

- Analyzes uploaded log files using file search + code interpreter.

- Can call 3 functions: `save_all_parameters`, `save_changed_parameters`, `save_minimal_parameters`.

- Fully manages assistant lifecycle:

- Reuses or creates assistant based on API key.

- Upgrade button replaces old assistant with latest version.

- Auto-updates assistant context after the user changes the updated logs file.

- Code interpreter filters and processes large datasets efficiently. Big potential for improvement here.

Known Issues:

- Sometimes the assistant wrongly thinks a log file was (or wasn’t) uploaded. Happens rarely and usually goes away after a page reload and context reset.

- Not tested on large or complex real-world log files, might struggle there.

- Occasionally fails to retrieve data due to schema mismatch, causing errors in code interpreter. Asking it to try again usually fixes it.

- Needs more testing to surface other edge cases.

### CODE REVIEW

File: HardwareReport/Chat.js [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+async function purgeOldSchemaFile(vectorStoreId, targetFilename = "logs_schema_and_notes.txt") {
+    const refs = await openai.beta.vectorStores.files.list(vectorStoreId);
+    if (!refs)
+        throw new Error("error retrieving vector store files list");
+    for (const ref of refs.data) {
+        const file = await openai.files.retrieve(ref.id);
+        if (!file)
+            throw new Error("error retrieving file");
+        if (file.filename === targetFilename) {
+        //detach from vector store
```
Comment (rmackay9) [inline HardwareReport/Chat.js:120]: nitpick: could indent the lines within the curly brackets

File: HardwareReport/Chat.js [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+        upgradeButton.title = 'Upgraded successfully to the newest Assistant version';
+        upgradeButton.textContent = 'Upgraded';
+    }
+        
+}
+
+//save the api key in the local storage, user will not have to re-enter it manually every time
+function saveAPIKey(){
+    const apiKey = document.getElementById('openai-api-key').value.trim();
+    localStorage.setItem('openai-api-key', apiKey);
```
[the remark this answers is not in the document]
Comment (AnasChebili) [reply HardwareReport/Chat.js]: Switched to sessionStorage, but since it's still vulnerable to XSS, we might want to look into adding a backend middleware for better security in the future.

**Comment by AnasChebili** [comment]
I've addressed all comments. Regarding the API key: it's now stored in memory instead of localStorage, it will no longer persist and the user will have to re-enter it when page is refreshed, this is slightly better for security but still carries XSS risks.

**Comment by rmackay9** [comment]
Hi @AnasChebili,

Thanks very much for the updates.  Feel free to mark the comments as resolved.  I know there's sometimes a question, "is it ok for the dev to mark them as resolved or should the reporter check and then mark them him/herself" but in general we let the dev do it.

Thanks!

**Comment by AnasChebili** [comment]
Hey, thanks for the feedback!
I fixed the API key issue. Now if someone tries to chat without a saved or valid key, it shows an assistant error in the chat instead of throwing alerts. Also made the "thinking" box auto-scroll properly.
The tool still works fine offline. I added a check for the CDN package on load, and if it’s missing, the assistant bubble stays hidden.
For API key persistence, we removed local storage earlier in this PR due to security concerns. It’s just kept in memory now. Still not perfect, so Randy and I talked about adding backend middleware to handle it more securely. Would love your thoughts on that.
As for the assistant not seeing the log, yeah, it sometimes hallucinates whether a file was uploaded. The file is getting passed correctly though. I updated the instruction set so it always tries to use it and only shows an error if it actually fails. That seems more reliable than pre-checking.
Let me know if anything else feels off
